File: 6_kyu/Find_the_unique_number.c
#include <assert.h>
#include <stddef.h>

// Returns the unique number in the array.
// All other numbers are equal.
float finduniq(const float nums[], const size_t n)
{
  assert( nums != NULL && n > 2 );
  // Start with estimation first number is the common number
  float common = nums[0];
  if( nums[1] != common )
  {
    // first or second number is not the common number
    if( nums[2] == common )
      return nums[1];
    else
      return nums[0];
  }
  // Now common is verified the common number.
  // Search the other number.
  for( size_t i = 2; i < n; ++i )
    if( nums[i] != common )
      return nums[i];
  // No unique number found
  return common;
}

File: 6_kyu/Find_the_unique_number.rs
// Returns the non-unique number in the array.
// All other numbers are equal. Only one number differs from the rest.
fn find_uniq(arr: &[f64]) -> f64 {
    if arr.len() < 3 {
        // not enough numbers to identify the "unique number"
        panic!();
    }
    // If the "unique number" is not at index 0 then it must at index 1.
    let the_number = if arr[0] == arr[1] || arr[0] == arr[2] {
        arr[0]
    } else {
        arr[1]
    };
    // Return the not unique number or panic.
    *(*arr).iter().find(|&x| *x != the_number).unwrap()
}
